### PR TITLE
directions-matrix

### PR DESCRIPTION
**Pull request by tsaifi9**

### CODE REVIEW

**Review by jclinford — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# go-mapbox-api
+
+API Wrapper for [Mapbox API](https://docs.mapbox.com/api/)
+
+## Usage
+
+### Initialization
+```go
+mapboxClient, err := NewClient(&MapboxConfig{
```
Comment (jclinford) [inline README.md:9]: We should move all of these examples to godoc so they don't become stale: https://godoc.org/github.com/fluhus/godoc-tricks#Examples

This readme should just be highlevel and link into godoc for any code or documentation

File: client.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	APIKey string
+}
+
+type Client struct {
+	httpClient *http.Client
+	apiKey string
+}
+
+// NewClient instantiates a new Mapbox client.
+func NewClient(config *MapboxConfig) (*Client, error) {
```
Comment (jclinford) [inline client.go:29]: !?!?!?!?!?

File: direction_matrix.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	Annotations Annotations
+	Approaches Approaches
+	Destinations Destinations
+	Sources Sources
+	FallbackSpeed FallbackSpeed
+
+}
+
+type DirectionsMatrixResponse struct {
+	Code string `json:"code"`
```
Comment (jclinford) [inline direction_matrix.go:29]: I don't think you ran gofmt on this code??

File: direction_matrix.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	Durations [][]float64 `json:"durations"`
+	Distances [][]float64 `json:"distances"`
+	Destinations []Waypoint `json:"destinations"`
+	Sources []Waypoint `json:"sources"`
+}
+
+// https://docs.mapbox.com/api/navigation/#matrix
+func (c *Client) DirectionsMatrix(ctx context.Context, req DirectionsMatrixRequest) (*DirectionsMatrixResponse, error) {
```
Comment (jclinford) [inline direction_matrix.go:37]: This is a little weird having the method of client not in the file with client.

Maybe have them in the same file, or something like this: ?

client.go
```go
func (c *Client) DirectionsMatrix(ctx context.Context, req DirectionsMatrixRequest) (*DirectionsMatrixResponse, error) {
    return directionsMatrix(ctx, c, req)
}
```

directions_matrix.go
```go
func directionsMatrix(ctx context.Context, c *Client, req DirectionsMatrixRequest) (*DirectionsMatrixResponse, error) {
    /// the implementation you have here
    /// .....
}
```

File: go.mod [hunk 5]
```diff
@@ -0,0 +1,3 @@
+module github.com/airspacetechnologies/go-mapbox
+
+go 1.12
```
Comment (jclinford) [inline go.mod:3]: You should be using go 1.13

File: types.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+const (
+	ProfileDriving = Profile("mapbox/driving")
+	ProfileWalking = Profile("mapbox/walking")
+	ProfileCycling = Profile("mapbox/cycling")
+	ProfileDrivingTraffic = Profile("mapbox/driving-traffic")
+
+	AnnotationDuration = Annotation("duration")
+	AnnotationDistance = Annotation("distance")
+	AnnotationSpeed = Annotation("speed")
+	AnnotationCongestion = Annotation("congestion")
```
Comment (jclinford) [inline types.go:17]: Nice, congestion! I think we could use this for anomalies!

File: direction_matrix.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	relPath := fmt.Sprintf("%v/%v/%v/%v", directionsMatrix, v1, req.Profile, req.Coordinates.WGS84Format())
+
+	query := url.Values{}
+	query.Set("access_token", c.apiKey)
+	query.Set("annotations", strings.Join(req.Annotations.strings(), ","))
```
Comment (jclinford) [inline direction_matrix.go:42]: Are things like annotations, etc used for other queries besides `DirectionsMatrix`?

If so we should move this join statement onto methods in `types.go`.

Either rename `strings()` to `string()` or have both if we need both. Or `asParam()` or something might also be a good name if you don't like the string names.

That way every endpoint doesn't have a bunch of strings.Join() calls that have to know if they join with semicolon or comma, etc. And every endpoint that we add will just call .string() on it.

File: types.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+	ApproachUnrestricted = Approach("unrestricted")
+	ApproachCurb = Approach("curb")
+)
+
+type Profile string
+
+type Annotations []Annotation
```
Comment (jclinford) [inline types.go:25]: gofmt your code!

File: README.md (hunk 1 above)
Comment (jclinford) [reply README.md:9]: We can wait to do this one until after we test out the API more

File: direction_matrix.go (hunk 7 above)
Comment (tsaifi9) [reply direction_matrix.go:42]: good callout

**Pull request merged**